File: Model/Actions/SoilActions.vb

Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.IO
Imports System.Net
Imports System.Reflection
Imports System.Text
Imports System.Windows.Forms
Imports System.Xml

Imports ApsimFile
Imports Controllers
Imports CSGeneral
Imports CSUserInterface 'SoilUI
Imports UIBits          'ErrorMessageForm
Imports UIUtility
Imports ExcelUtility


Public Class SoilActions

    Public Shared Sub FileNew(ByVal Controller As BaseController)
        If Controller.FileSaveAfterPrompt() Then
            Dim Dialog As New SaveFileDialog()
            Dialog.Filter = "Soils files (*.soils)|*.soils|" + "All files (*.*)|*.*"
            Dialog.DefaultExt = "soils"
            Dialog.Title = "Provide a filename to save the new soils file to"
            If Dialog.ShowDialog() = DialogResult.OK Then
                Dim Out As New StreamWriter(Dialog.FileName)
                Out.WriteLine("<folder name=""Soils"" version=""" + ApsimFile.APSIMChangeTool.CurrentVersion.ToString() + """/>")
                Out.Close()
                Controller.ApsimData.OpenFile(Dialog.FileName)
            End If
        End If
    End Sub
   Public Shared Sub AddSoil(ByVal Controller As BaseController)
      Dim NewSoil As XmlNode = Soil.Create("Soil")
      Controller.Selection.Add(NewSoil.OuterXml)
   End Sub




#Region "Import methods"
    Public Shared Sub ImportFromSoils(ByVal Controller As BaseController)
        Dim Dialog As New OpenFileDialog()
        Dialog.Filter = "Soils files (*.soils)|*.soils|All files (*.*)|*.*"
        Dialog.Title = "Select 1 or more .soils file(s) to import from"
        Dialog.Multiselect = True
        Dialog.RestoreDirectory = True
        If Dialog.ShowDialog() = DialogResult.OK Then
            For Each FileName As String In Dialog.FileNames
                Dim Doc As New XmlDocument()
                Doc.Load(FileName)
                APSIMChangeTool.Upgrade(Doc.DocumentElement)
                Controller.Selection.Add(Doc.DocumentElement.InnerXml)
            Next
        End If
    End Sub
    Public Shared Sub ImportFromPar(ByVal Controller As BaseController)
        Dim Dialog As New OpenFileDialog()
        Dialog.Filter = "Par files (*.par)|*.par|All files (*.*)|*.*"
        Dialog.Title = "Select 1 or more .par file(s) to import from"
        Dialog.Multiselect = True
        Dialog.RestoreDirectory = True
        If Dialog.ShowDialog() = DialogResult.OK Then
            For Each FileName As String In Dialog.FileNames
                Cursor.Current = Cursors.WaitCursor
                Controller.Selection.Add(SoilParFileImporter.Import(FileName))
                Cursor.Current = Cursors.Default
            Next
        End If
   End Sub

   Private Shared ProgressBar As ToolStripProgressBar

   Public Shared Sub FileOpenXLS(ByVal Controller As BaseController)
      Dim StatusBar As StatusStrip = Controller.MainForm.Controls.Find("StatusStrip1", True)(0)
      ProgressBar = StatusBar.Items(0)
      StatusBar.Visible = True

      Dim Dialog As New OpenFileDialog()
      Dialog.Filter = "Spreadsheet files (*.xls)|*.xls|All files (*.*)|*.*"
      Dialog.Title = "Select a spreadsheet to open"
      Dialog.RestoreDirectory = True
      If Dialog.ShowDialog() = DialogResult.OK Then
         Cursor.Current = Cursors.WaitCursor
         Dim TempFileName As String = SoilSpreadsheet.OpenXLS(Dialog.FileName, AddressOf UpdateProgress)
         Controller.ApsimData.NewFromFile(TempFileName)
         File.Delete(TempFileName)
         Cursor.Current = Cursors.Default
      End If
      StatusBar.Visible = False
   End Sub

   Private Shared Sub UpdateProgress(ByVal Percent As Integer)
      ProgressBar.Value = Percent
      Application.DoEvents()
   End Sub

#End Region



#Region "Export methods"

    Public Shared Sub ExportToSoils(ByVal Controller As BaseController)
        Dim Dialog As New SaveFileDialog()
        Dialog.Filter = "Soils files (*.soils)|*.soils|All files (*.*)|*.*"
        Dialog.Title = "Enter a .soils file to export to"
        Dialog.DefaultExt = "soils"
        If Dialog.ShowDialog() = DialogResult.OK Then
            Dim Doc As New XmlDocument()
            If Not File.Exists(Dialog.FileName) Then
                Doc.AppendChild(XmlHelper.CreateNode(Doc, "soils", ""))
            Else
                Doc.Load(Dialog.FileName)
            End If

            For Each SelectedPath As String In Controller.SelectedPaths
                Dim Comp As ApsimFile.Component = Controller.ApsimData.Find(SelectedPath)
                Dim NodeDoc As New XmlDocument()
            NodeDoc.LoadXml(Comp.FullXML())
                Doc.DocumentElement.AppendChild(Doc.ImportNode(NodeDoc.DocumentElement, True))
            Next
            XmlHelper.SetAttribute(Doc.DocumentElement, "version", ApsimFile.APSIMChangeTool.CurrentVersion.ToString())
            Doc.Save(Dialog.FileName)
            MessageBox.Show("Soils have been successfully exported to '" + Dialog.FileName + "'. It is suggested that you rename soils within the new file to avoid confusion.", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information)
        End If
    End Sub
   Public Shared Sub ExportToSpreadsheet(ByVal Controller As BaseController)
      Dim Dialog As New SaveFileDialog()
      Dialog.Filter = "Spreadsheet files (*.xls)|*.xls|All files (*.*)|*.*"
      Dialog.Title = "Enter a spreadsheet file to export to"
      Dialog.DefaultExt = "xls"
      If Dialog.ShowDialog() = DialogResult.OK Then
         Dim Doc As New XmlDocument
         Doc.LoadXml(Controller.ApsimData.RootComponent.FullXML())
         SoilSpreadsheet.Export(Dialog.FileName, Doc.DocumentElement, Controller.SelectedPaths)
         MessageBox.Show("Soils have been successfully exported to '" + Dialog.FileName + "'. It is suggested that you rename soils within the new file to avoid confusion.", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information)
      End If
   End Sub

#End Region



    Public Shared Sub CheckSoils(ByVal Controller As BaseController)
        ' User wants to check all soils for consistency 
        Cursor.Current = Cursors.WaitCursor
        Dim ErrorMessage As String = ""
        For Each SelectedPath As String In Controller.SelectedPaths
            CheckSoils(Controller.ApsimData.Find(SelectedPath), ErrorMessage)
        Next
        If ErrorMessage = "" Then
            MessageBox.Show("All soils checked out ok. No problems were encountered", "No problems encountered", MessageBoxButtons.OK, MessageBoxIcon.Information)
        Else
            Dim ErrorForm As New UIBits.ErrorMessageForm()
            ErrorForm.SetText(ErrorMessage)
            ErrorForm.Show()
        End If
        Cursor.Current = Cursors.[Default]
    End Sub
    Private Shared Sub CheckSoils(ByVal Data As ApsimFile.Component, ByRef ErrorMessage As String)
        If Data.Type.ToLower() = "soil" Then
         Dim ThisSoil As XmlNode = Soil.CreateFromXML(Data.FullXML())
         Dim Errors As String = Soil.CheckForErrors(ThisSoil, True)
            If Errors <> "" Then
            ErrorMessage += vbCr + vbLf + XmlHelper.Name(ThisSoil) + vbCr + vbLf + StringManip.IndentText(Errors, 6)
            End If
        ElseIf Data.Type.ToLower() = "folder" Then
            For Each Child As ApsimFile.Component In Data.ChildNodes
                CheckSoils(Child, ErrorMessage)
            Next
        End If
    End Sub

    Public Shared Sub SortSoils(ByVal Controller As BaseController)
        Cursor.Current = Cursors.WaitCursor
        Controller.Selection.Sort()
        Cursor.Current = Cursors.[Default]
    End Sub


    Public Shared Sub Version(ByVal Controller As BaseController)
        MessageBox.Show(VersionString(), "Apsoil version", MessageBoxButtons.OK, MessageBoxIcon.Information)
    End Sub
    Public Shared Function VersionString() As String
        Return "Version " + Configuration.Instance.ApsimVersion()
    End Function
   Public Shared Sub OpenLatestVersionOfSoilsDatabase(ByVal Controller As BaseController)
      If Controller.FileSaveAfterPrompt() Then

         Cursor.Current = Cursors.WaitCursor
         Dim request As WebRequest = WebRequest.Create("http://www.apsim.info/Wiki/public/Upload/ApSoil/APSRU-Australia-Soils.soils")
         Dim response As HttpWebResponse = DirectCast(request.GetResponse(), HttpWebResponse)
         If response.StatusDescription = "OK" Then
            Dim dataStream As Stream = response.GetResponseStream()
            Dim reader As New StreamReader(dataStream)
            Dim responseFromServer As String = reader.ReadToEnd()

            Dim SoilsFileName As String = Path.GetTempFileName()
            Dim SoilsFile As New StreamWriter(SoilsFileName)
            SoilsFile.Write(responseFromServer)
            SoilsFile.Close()
            reader.Close()
            dataStream.Close()
            Controller.ApsimData.NewFromFile(SoilsFileName)
            File.Delete(SoilsFileName)
         Else
            MessageBox.Show("Cannot connect to www.apsim.info", "Failure", MessageBoxButtons.OK, MessageBoxIcon.[Error])
         End If
         response.Close()
         Cursor.Current = Cursors.[Default]
      End If
   End Sub
    Public Shared Sub ReleaseNotes(ByVal Controller As BaseController)
        Dim URL As String = Configuration.Instance.Setting("ReleaseNotes")
        System.Diagnostics.Process.Start(URL)
    End Sub

   Public Shared Sub GoogleEarthSoils(ByVal Controller As BaseController)
      If Controller.FileSaveAfterPrompt() Then
         Cursor.Current = Cursors.WaitCursor
         Dim request As WebRequest = WebRequest.Create("http://www.apsim.info/ApsoilWeb/ApsoilKML.aspx")
         Dim response As HttpWebResponse = DirectCast(request.GetResponse(), HttpWebResponse)
         If response.StatusDescription = "OK" Then
            Dim dataStream As Stream = response.GetResponseStream()
            Dim reader As New StreamReader(dataStream)
            Dim responseFromServer As String = reader.ReadToEnd()

            Dim KMLFileName As String = Path.GetTempPath() + "Soils.kmz"
            Dim KMLFile As New StreamWriter(KMLFileName)
            KMLFile.Write(responseFromServer)
            KMLFile.Close()
            reader.Close()
            dataStream.Close()
            Dim P As Process = CSGeneral.Utility.RunProcess(KMLFileName, "", Path.GetTempPath())
            Dim Errors As String = CSGeneral.Utility.CheckProcessExitedProperly(P)
            If Errors <> "" Then
               MessageBox.Show(Errors, "Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
            End If

         Else
            MessageBox.Show("Cannot connect to www.apsim.info", "Failure", MessageBoxButtons.OK, MessageBoxIcon.[Error])
         End If
         response.Close()
         Cursor.Current = Cursors.[Default]
      End If
   End Sub



#Region "PrintSoil"
    Public Shared Sub PrintSoil(ByVal Controller As BaseController)
        'Dim SoilUI As CSUserInterface.SoilUI = DirectCast(Controller.Explorer.CurrentView, CSUserInterface.SoilUI)

        'Dim PrintDocument As New System.Drawing.Printing.PrintDocument()
        'AddHandler PrintDocument.BeginPrint, AddressOf SoilUI.OnBeginPrint
        'AddHandler PrintDocument.PrintPage, AddressOf SoilUI.OnPrintPage
        'AddHandler PrintDocument.QueryPageSettings, AddressOf OnQueryPageSettings

        'Dim PrintDialog As New PrintDialog()
        'PrintDialog.Document = PrintDocument

        'Dim PreviewDialog As New System.Windows.Forms.PrintPreviewDialog()
        'PreviewDialog.Document = PrintDocument
        'PreviewDialog.ShowDialog()
    End Sub
    Private Shared Sub OnQueryPageSettings(ByVal sender As Object, ByVal e As System.Drawing.Printing.QueryPageSettingsEventArgs)
        e.PageSettings.Margins.Left = 50
        e.PageSettings.Margins.Top = 50
        e.PageSettings.Margins.Right = 50
        e.PageSettings.Margins.Bottom = 50
    End Sub
#End Region


End Class

File: Model/Actions/SoilActions.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;

using System.Drawing;
using System.IO;
using System.Net;
using System.Reflection;
using System.Text;
using System.Windows.Forms;
using System.Xml;

using ApsimFile;
using Controllers;
using CSGeneral;
using CSUserInterface;
//SoilUI
using UIBits;
//ErrorMessageForm
using UIUtility;
using ExcelUtility;
namespace Actions
{


	public class SoilActions
	{

		public static void FileNew(BaseController Controller)
		{
			if (Controller.FileSaveAfterPrompt()) {
				SaveFileDialog Dialog = new SaveFileDialog();
				Dialog.Filter = "Soils files (*.soils)|*.soils|" + "All files (*.*)|*.*";
				Dialog.DefaultExt = "soils";
				Dialog.Title = "Provide a filename to save the new soils file to";
				if (Dialog.ShowDialog() == DialogResult.OK) {
					StreamWriter Out = new StreamWriter(Dialog.FileName);
					Out.WriteLine("<folder name=\"Soils\" version=\"" + ApsimFile.APSIMChangeTool.CurrentVersion.ToString() + "\"/>");
					Out.Close();
					Controller.ApsimData.OpenFile(Dialog.FileName);
				}
			}
		}
		public static void AddSoil(BaseController Controller)
		{
            throw new NotImplementedException();
            //XmlNode NewSoil = Soil.Create("Soil");
            //Controller.Selection.Add(NewSoil.OuterXml);
		}




		#region "Import methods"
		public static void ImportFromSoils(BaseController Controller)
		{
			OpenFileDialog Dialog = new OpenFileDialog();
			Dialog.Filter = "Soils files (*.soils)|*.soils|All files (*.*)|*.*";
			Dialog.Title = "Select 1 or more .soils file(s) to import from";
			Dialog.Multiselect = true;
			Dialog.RestoreDirectory = true;
			if (Dialog.ShowDialog() == DialogResult.OK) {
				foreach (string FileName in Dialog.FileNames) {
					XmlDocument Doc = new XmlDocument();
					Doc.Load(FileName);
					APSIMChangeTool.Upgrade(Doc.DocumentElement);
					Controller.Selection.Add(Doc.DocumentElement.InnerXml);
				}
			}
		}
		public static void ImportFromPar(BaseController Controller)
		{
			OpenFileDialog Dialog = new OpenFileDialog();
			Dialog.Filter = "Par files (*.par)|*.par|All files (*.*)|*.*";
			Dialog.Title = "Select 1 or more .par file(s) to import from";
			Dialog.Multiselect = true;
			Dialog.RestoreDirectory = true;
			if (Dialog.ShowDialog() == DialogResult.OK) {
				foreach (string FileName in Dialog.FileNames) {
					Cursor.Current = Cursors.WaitCursor;
					Controller.Selection.Add(SoilParFile.Import(FileName));
					Cursor.Current = Cursors.Default;
				}
			}
		}


		private static ToolStripProgressBar ProgressBar;
		public static void FileOpenXLS(BaseController Controller)
		{
			StatusStrip StatusBar = (StatusStrip)Controller.MainForm.Controls.Find("StatusStrip1", true)[0];
			ProgressBar = (ToolStripProgressBar)StatusBar.Items[0];
			StatusBar.Visible = true;

			OpenFileDialog Dialog = new OpenFileDialog();
			Dialog.Filter = "Spreadsheet files (*.xls;*.xlsx)|*.xls;*.xlsx|All files (*.*)|*.*";
			Dialog.Title = "Select a spreadsheet to open";
			Dialog.RestoreDirectory = true;
			if (Dialog.ShowDialog() == DialogResult.OK) {
				Cursor.Current = Cursors.WaitCursor;
				DataTable Table = ExcelUtility.ExcelHelper.GetDataFromSheet(Dialog.FileName, "SoilData");
                XmlNode Data = SoilDataTable.TableToSoilXML(Table, UpdateProgress);
				Controller.ApsimData.New(Data.OuterXml);
				Cursor.Current = Cursors.Default;
			}
			StatusBar.Visible = false;
		}

		private static void UpdateProgress(int Percent)
		{
			ProgressBar.Value = Percent;
			Application.DoEvents();
		}

		#endregion



		#region "Export methods"

		public static void ExportToSoils(BaseController Controller)
		{
			SaveFileDialog Dialog = new SaveFileDialog();
			Dialog.Filter = "Soils files (*.soils)|*.soils|All files (*.*)|*.*";
			Dialog.Title = "Enter a .soils file to export to";
			Dialog.DefaultExt = "soils";
			if (Dialog.ShowDialog() == DialogResult.OK) {
				XmlDocument Doc = new XmlDocument();
				if (!File.Exists(Dialog.FileName)) {
					Doc.AppendChild(XmlHelper.CreateNode(Doc, "soils", ""));
				} else {
					Doc.Load(Dialog.FileName);
				}

				foreach (string SelectedPath in Controller.SelectedPaths) {
					ApsimFile.Component Comp = Controller.ApsimData.Find(SelectedPath);
					XmlDocument NodeDoc = new XmlDocument();
					NodeDoc.LoadXml(Comp.FullXML());
					Doc.DocumentElement.AppendChild(Doc.ImportNode(NodeDoc.DocumentElement, true));
				}
				XmlHelper.SetAttribute(Doc.DocumentElement, "version", ApsimFile.APSIMChangeTool.CurrentVersion.ToString());
				Doc.Save(Dialog.FileName);
				MessageBox.Show("Soils have been successfully exported to '" + Dialog.FileName + "'. It is suggested that you rename soils within the new file to avoid confusion.", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information);
			}
		}
		public static void ExportToSpreadsheet(BaseController Controller)
		{
			SaveFileDialog Dialog = new SaveFileDialog();
			Dialog.Filter = "Spreadsheet files (*.xlsx)|*.xlsx|All files (*.*)|*.*";
			Dialog.Title = "Enter a spreadsheet file to export to";
			Dialog.DefaultExt = "xlsx";
            Dialog.OverwritePrompt = true;
			if (Dialog.ShowDialog() == DialogResult.OK) {
                Cursor.Current = Cursors.WaitCursor;
				XmlDocument Doc = new XmlDocument();
				Doc.LoadXml(Controller.ApsimData.RootComponent.FullXML());
				List<string> Paths = new List<string>();
				foreach (string Path in Controller.SelectedPaths) {
					Paths.Add(Path);
				}
                DataTable Table = SoilDataTable.SoilXMLToTable(Doc.DocumentElement, Paths);
				ExcelHelper.SendDataToSheet(Dialog.FileName, "SoilData", Table);
                Cursor.Current = Cursors.Default;
				MessageBox.Show("Soils have been successfully exported to '" + Dialog.FileName + "'. It is suggested that you rename soils within the new file to avoid confusion.", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information);
			}
		}

		#endregion



		public static void CheckSoils(BaseController Controller)
		{
			// User wants to check all soils for consistency 
			Cursor.Current = Cursors.WaitCursor;
			string ErrorMessage = "";
			foreach (string SelectedPath in Controller.SelectedPaths) {
				CheckSoils(Controller.ApsimData.Find(SelectedPath), ref ErrorMessage);
			}
			if (string.IsNullOrEmpty(ErrorMessage)) {
				MessageBox.Show("All soils checked out ok. No problems were encountered", "No problems encountered", MessageBoxButtons.OK, MessageBoxIcon.Information);
			} else {
				UIBits.ErrorMessageForm ErrorForm = new UIBits.ErrorMessageForm();
				ErrorForm.SetText(ErrorMessage);
				ErrorForm.Show();
			}
			Cursor.Current = Cursors.Default;
		}
		private static void CheckSoils(ApsimFile.Component Data, ref string ErrorMessage)
		{
			if (Data.Type.ToLower() == "soil") {
                throw new NotImplementedException();
                //XmlNode ThisSoil = Soil.CreateFromXML(Data.FullXML());
				//string Errors = Soil.CheckForErrors(ThisSoil, true);
                //if (!string.IsNullOrEmpty(Errors)) {
                //    ErrorMessage += Environment.NewLine + XmlHelper.Name(ThisSoil) + Environment.NewLine + StringManip.IndentText(Errors, 6);
				//}
			} else if (Data.Type.ToLower() == "folder") {
				foreach (ApsimFile.Component Child in Data.ChildNodes) {
					CheckSoils(Child, ref ErrorMessage);
				}
			}
		}

		public static void SortSoils(BaseController Controller)
		{
			Cursor.Current = Cursors.WaitCursor;
			Controller.Selection.Sort();
			Cursor.Current = Cursors.Default;
		}


		public static void Version(BaseController Controller)
		{
			MessageBox.Show(VersionString(), "Apsoil version", MessageBoxButtons.OK, MessageBoxIcon.Information);
		}
		public static string VersionString()
		{
			return "Version " + Configuration.Instance.Setting("Version");
		}
		public static void OpenLatestVersionOfSoilsDatabase(BaseController Controller)
		{

			if (Controller.FileSaveAfterPrompt()) {
				Cursor.Current = Cursors.WaitCursor;
				WebRequest request = WebRequest.Create("http://www.apsim.info/Wiki/public/Upload/ApSoil/APSRU-Australia-Soils.soils");
				HttpWebResponse response = (HttpWebResponse)request.GetResponse();
				if (response.StatusDescription == "OK") {
					Stream dataStream = response.GetResponseStream();
					StreamReader reader = new StreamReader(dataStream);
					string responseFromServer = reader.ReadToEnd();

					string SoilsFileName = Path.GetTempFileName();
					StreamWriter SoilsFile = new StreamWriter(SoilsFileName);
					SoilsFile.Write(responseFromServer);
					SoilsFile.Close();
					reader.Close();
					dataStream.Close();
					Controller.ApsimData.NewFromFile(SoilsFileName);
					File.Delete(SoilsFileName);
				} else {
					MessageBox.Show("Cannot connect to www.apsim.info", "Failure", MessageBoxButtons.OK, MessageBoxIcon.Error);
				}
				response.Close();
				Cursor.Current = Cursors.Default;
			}
		}
		public static void ReleaseNotes(BaseController Controller)
		{
			string URL = Configuration.Instance.Setting("ReleaseNotes");
			System.Diagnostics.Process.Start(URL);
		}

		public static void GoogleEarthSoils(BaseController Controller)
		{
			if (Controller.FileSaveAfterPrompt()) {
				Cursor.Current = Cursors.WaitCursor;
				WebRequest request = WebRequest.Create("http://www.apsim.info/ApsoilWeb/ApsoilKML.aspx");
				HttpWebResponse response = (HttpWebResponse)request.GetResponse();
				if (response.StatusDescription == "OK") {
					Stream dataStream = response.GetResponseStream();
					StreamReader reader = new StreamReader(dataStream);
					string responseFromServer = reader.ReadToEnd();

					string KMLFileName = Path.GetTempPath() + "Soils.kmz";
					StreamWriter KMLFile = new StreamWriter(KMLFileName);
					KMLFile.Write(responseFromServer);
					KMLFile.Close();
					reader.Close();
					dataStream.Close();
					Process P = CSGeneral.Utility.RunProcess(KMLFileName, "", Path.GetTempPath());
					string Errors = CSGeneral.Utility.CheckProcessExitedProperly(P);
					if (!string.IsNullOrEmpty(Errors)) {
						MessageBox.Show(Errors, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
					}

				} else {
					MessageBox.Show("Cannot connect to www.apsim.info", "Failure", MessageBoxButtons.OK, MessageBoxIcon.Error);
				}
				response.Close();
				Cursor.Current = Cursors.Default;
			}
		}



		#region "PrintSoil"
		public static void PrintSoil(BaseController Controller)
		{
			//Dim SoilUI As CSUserInterface.SoilUI = DirectCast(Controller.Explorer.CurrentView, CSUserInterface.SoilUI)

			//Dim PrintDocument As New System.Drawing.Printing.PrintDocument()
			//AddHandler PrintDocument.BeginPrint, AddressOf SoilUI.OnBeginPrint
			//AddHandler PrintDocument.PrintPage, AddressOf SoilUI.OnPrintPage
			//AddHandler PrintDocument.QueryPageSettings, AddressOf OnQueryPageSettings

			//Dim PrintDialog As New PrintDialog()
			//PrintDialog.Document = PrintDocument

			//Dim PreviewDialog As New System.Windows.Forms.PrintPreviewDialog()
			//PreviewDialog.Document = PrintDocument
			//PreviewDialog.ShowDialog()
		}
		private static void OnQueryPageSettings(object sender, System.Drawing.Printing.QueryPageSettingsEventArgs e)
		{
			e.PageSettings.Margins.Left = 50;
			e.PageSettings.Margins.Top = 50;
			e.PageSettings.Margins.Right = 50;
			e.PageSettings.Margins.Bottom = 50;
		}
		#endregion


	}
}




